test_apps/smokeTests/gridTools.cpp

```diff
@@ -0,0 +1,594 @@
+    double sum     = 0;
+    for ( size_t k=0; k<z; k++ ) {
+        for ( size_t j=0; j<y; j++ ) {
+            for ( size_t i=0; i<x; i++ ) {
+                size_t indices[3] = { i, j, k };
+                //float trueValue = grid.AccessIJK( i, j, k );
+                double trueValue = grid->GetValueAtIndex( {i,j,k} );
+                
+                double coords[3];
+                grid->GetUserCoordinates( indices, coords );
+                float sampleValue = grid->GetValue( coords );
+                
+                if (sampleValue == grid->GetMissingValue()) {
+                    numMissingValues++;
+                    continue;
+                }
+
+                double error = abs( sampleValue - trueValue );
+
+                if (error != 0) {
+                    disagreements++;
+                }
+
+                if ( error > peak ) 
+                    peak = error;
+                sum += error*error;
+
+            }
+        }
+    }
+
+    rms = sqrt( sum / (x*y*z) );
+
+    if ( rms != 0 || disagreements > 0 )
+        return 1;
+    return 0;
+}
+
+bool isNotEqual(
+    double x, 
+    double y
+) {
+  const double epsilon = 1e-5;
+  return std::abs(x - y) > epsilon * std::abs(x);
+}
+
+size_t TestConstNodeIterator( 
+    const Grid *g, 
+    size_t& count, 
+    size_t& disagreements, 
+    double& time 
+) {
+    Grid::ConstNodeIterator itr;
+    Grid::ConstNodeIterator enditr = g->ConstNodeEnd();
+
+    double t0 = Wasp::GetTime();
+    itr = g->ConstNodeBegin();
+    
+    std::vector< size_t > dims = g->GetDimensions();
+
+    for ( ; itr!=enditr; ++itr) {
+        size_t i = count % dims[X];
+        size_t j = ( count / dims[X] ) % dims[Y];
+        size_t k = count / ( dims[X] * dims[Y] );
+       
+        double itrData  = g->GetValueAtIndex( (*itr).data() );
+        double gridData = g->AccessIJK( i, j, k ); 
+        if ( isNotEqual( itrData, gridData ) )
+            disagreements++;
+
+        count++;
+    }
+
+    size_t expectedCount = 1;
+    for ( auto dim : dims )
+        expectedCount *= dim;
+
+    time = Wasp::GetTime() - t0;
+
+    return expectedCount;
+}
+
+size_t TestIterator( 
+    Grid *g, 
+    size_t& count, 
+    size_t& disagreements,
+    double& time 
+) {
+    Grid::Iterator itr;
+    Grid::Iterator enditr = g->end();
+
+    double t0 = Wasp::GetTime();
+    itr = g->begin();
+    
+    std::vector<size_t> dims = g->GetDimensions();
+
+    for ( ; itr!=enditr; ++itr) {
+        size_t i = count % dims[X];
+        size_t j = ( count / dims[X] ) % dims[Y];
+        //size_t j = count / dims[X];
+        size_t k = count / ( dims[X] * dims[Y] );
+        
+        if ( isNotEqual( *itr, g->AccessIJK( i, j, k) ) )
+            disagreements++;
+
+        count++;
+    }
+
+    size_t expectedCount = 1;
+    for ( auto dim : dims )
+        expectedCount *= dim;
+
+    time = Wasp::GetTime() - t0;
+
+    return expectedCount;
+}
+
+size_t TestConstCoordItr( 
```

Comment:
Same comments as for TestConstNodeIterator

Reply:
Fixed